### PR TITLE
Add custom formatters

### PR DESCRIPTION
**Pull request by denisart**
Hooks in other libs: https://the-guild.dev/graphql/codegen/docs/config-reference/lifecycle-hooks

it is beta version of custom code formatters. it is simple realization without lifecycle hooks. Only before code generation

See `docs/custom-formatters.md` for simple usage example. And see `test_format.py` for development examples.

What do you think about this idea?

### CODE REVIEW

**Comment by koxudaxi** [comment]
@denisart 
Thank you for creating the PR.
We need the feature to use another formatter. We have the feature request. https://github.com/koxudaxi/datamodel-code-generator/issues/621

This PR provides a flexible way to pass the formatter.
But, can we give the preset formatter like ruff? https://github.com/koxudaxi/datamodel-code-generator/issues/1643
(I want to change the default formatter to ruff. Because black spend time in the process.)
image https://codspeed.io/koxudaxi/datamodel-code-generator/branches/denisart:add-custom-formatters


I expect the CLI arguments will be added in the future.
```
--preset-formatter-python [ruff, black-isort]
or 
--preset-formatter [ruff, black-isort, gofmt]
```

File: datamodel_code_generator/format.py [hunk 1]
```diff
@@ -167,12 +170,49 @@ def __init__(
...
     def format_code(
         self,
         code: str,
     ) -> str:
         code = self.apply_isort(code)
         code = self.apply_black(code)
+
+        for formatter in self.custom_formatters:
+            code = formatter.apply(code)
+
```
Comment (koxudaxi) [inline datamodel_code_generator/format.py:215]: Should the method apply all formatter?
Someone may want to replace black/isort formatters with others like ruff.

Comment (denisart) [reply datamodel_code_generator/format.py:215]: > Should the method apply all formatter?

I also agree. I think that in the future we need to add popular custom formatters and freedom of choice.

I think that this method should apply the list of formatters specified by the user like this
```bash
datamodel-codegen ... \
  --formatters \
  "datamodel_code_generator.ruff,user_formatter_module"
```

**Comment by denisart** [comment]
> But, can we give the preset formatter like ruff?

I agree with the ruff as default (if user not specified formatters). Do you want to work on this PR or as part of a new PR?

File: datamodel_code_generator/format.py (hunk 1 above)
Comment (denisart) [reply datamodel_code_generator/format.py:215]: > ... apply the list of formatters specified by the user ...

or default formatter (ruff) or disable format (special flag)

Comment (koxudaxi) [reply datamodel_code_generator/format.py:215]: OK, What do you think about the option?
```
--default-formatter  [black-isort(default), ruff, none]
```` 
`none` maybe not clearly?
Or
```
--default-formatter  [black-isort(default), ruff]
and 
--disable-default-formatter 
````

**Comment by koxudaxi** [comment]
@denisart 
> I agree with the ruff as default (if user not specified formatters). Do you want to work on this PR or as part of a new PR?

Thank you for asking.
I don't expect to set `ruff` as default soon. Because I guess the changes will affect many users.
But, I prefer to introduce `ruff` option as preset.

This project is used by many projects. I want to avoid breaking changes in the short term. (including adding dependency )

File: datamodel_code_generator/format.py (hunk 1 above)
Comment (denisart) [reply datamodel_code_generator/format.py:215]: The second option seems clearer to me.

Then we have 4 args:

`--default-formatter`: choice one (or a few ?) default formatters;
`--disable-default-formatter`: disable default formatters
`--custom-formatters`: user formatters
`--custom-formatters-kwargs`: json file with params for user formatters

Comment (koxudaxi) [reply datamodel_code_generator/format.py:215]: Looks good.
We have discussed generating non-python code by the CLI.
We should accept another language formatter or turn off the default formatter.
https://github.com/koxudaxi/datamodel-code-generator/discussions/1657

Comment (denisart) [reply datamodel_code_generator/format.py:215]: > generating non-python code by the CLI

It is good idea! I have a few questions about this.
I will take a look and answer in the discussion.

**Comment by denisart** [comment]
> I want to avoid breaking changes in the short term.

I agree with you.

File: datamodel_code_generator/format.py (hunk 1 above)
Comment (denisart) [reply datamodel_code_generator/format.py:215]: @koxudaxi 

I answer here. I agree with [#1657](https://github.com/koxudaxi/datamodel-code-generator/discussions/1657).
I'm thinking about a more universal approach myself. I have some thoughts about realization details.
I'll try to move in this direction in the next commits.

Comment (denisart) [reply datamodel_code_generator/format.py:215]: Maybe we can create a discussion where we discuss "road map" for `datamodel-code-generator` version `1.0.0`?

Comment (koxudaxi) [reply datamodel_code_generator/format.py:215]: I just created it.
https://github.com/koxudaxi/datamodel-code-generator/discussions/1736

**Comment by denisart** [comment]
@koxudaxi 
Do we have any undiscussed issues about this PR?

**Review by koxudaxi — APPROVED** [review]
LGTM.

**Pull request merged**